raphtory/src/search/graph_index.rs

```diff
@@ -360,11 +213,251 @@ impl GraphIndex {
         layer: usize,
         props: &[(usize, Prop)],
     ) -> Result<(), GraphError> {
-        self.edge_index
+        self.index
+            .edge_index
             .update_edge_constant_properties(edge_id, layer, props)
     }
 }
 
+#[derive(Clone)]
+pub enum GraphIndex {
+    Immutable(ImmutableGraphIndex),
+    Mutable(MutableGraphIndex),
+}
+
+impl Debug for GraphIndex {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            GraphIndex::Immutable(i) => f
+                .debug_struct("GraphIndex::Immutable")
+                .field("node_index", &i.index.node_index)
+                .field("edge_index", &i.index.edge_index)
+                .field("path", &i.path)
+                .field("index_spec", &i.index_spec)
+                .finish(),
+
+            GraphIndex::Mutable(m) => f
+                .debug_struct("GraphIndex::Mutable")
+                .field("node_index", &m.index.node_index)
+                .field("edge_index", &m.index.edge_index)
+                .field("path", &m.path.as_ref().map(|p| p.path()))
+                .field("index_spec", &m.index_spec)
+                .finish(),
+        }
+    }
+}
+
+impl GraphIndex {
+    pub fn create(
+        graph: &GraphStorage,
+        create_in_ram: bool,
+        cached_graph_path: Option<GraphFolder>,
+    ) -> Result<Self, GraphError> {
+        let dir = if !create_in_ram {
+            let temp_dir = match cached_graph_path {
+                // Creates index in a temp dir within cache graph dir.
+                // The intention is to avoid creating index in a tmp dir that could be on another file system.
+                Some(path) => TempDir::new_in(path.get_base_path())?,
+                None => TempDir::new()?,
+            };
+
+            Some(Arc::new(temp_dir))
+        } else {
+            None
+        };
+
+        let index_spec = IndexSpec::default();
+
+        let path = get_node_index_path(&dir);
+        let node_index = NodeIndex::new(&path)?;
+        node_index.index_nodes(graph, path, &index_spec)?;
+
+        let path = get_edge_index_path(&dir);
+        let edge_index = EdgeIndex::new(&path)?;
+        edge_index.index_edges(graph, path, &index_spec)?;
+
+        Ok(GraphIndex::Mutable(MutableGraphIndex {
+            index: Index {
+                node_index,
+                edge_index,
+            },
+            path: dir,
+            index_spec: Arc::new(RwLock::new(index_spec)),
+        }))
+    }
+
+    pub fn load_from_path(path: &GraphFolder) -> Result<GraphIndex, GraphError> {
+        let tmp_path = TempDir::new()?;
+        let mut index_path = path.get_index_path();
+        if path.prefer_zip_format {
+            index_path = tmp_path.keep();
+            unzip(&path.get_base_path(), index_path.as_path())?
+        };
```

Comment:
Several problems with this:
1. don't rely on `path.prefer_zip_format` for loading, there is no guarantee that it matches what is on disk
2. TempDir should only be created if we are actually loading from zip
3. If we are loading from zip and made a TempDir, we should return a mutable index as we have already copied the data!